### PR TITLE
feat: select year screen

### PR DESCRIPTION
**Pull request by franciscog-kainos**
### JIRA link
https://companieshouse.atlassian.net/browse/LFA-1795

### Change description
Added select year screen
image

image

### Work checklist

- [x] Tests added where applicable
- [x] UI changes look good on mobile
- [ ] UI changes meet accessibility criteria

### Merge instructions

We are commited to keeping commit history clean, consistent and linear. To achive this commit should be structured as follows:

```
<type>[optional scope]: <description>
```

and contain the following structural elements:

- fix: a commit that patches a bug in your codebase (this correlates with PATCH in semantic versioning),
- feat: a commit that introduces a new feature to the codebase (this correlates with MINOR in semantic versioning),
- BREAKING CHANGE: a commit that has a footer `BREAKING CHANGE:` introduces a breaking API change (correlating with MAJOR in semantic versioning). A BREAKING CHANGE can be part of commits of any type,
- types other than `fix:` and `feat:` are allowed, for example `build:`, `chore:`, `ci:`, `docs:`, `style:`, `refactor:`, `perf:`, `test:`, and others,
- footers other than `BREAKING CHANGE: <description>` may be provided.

### CODE REVIEW

File: src/views/select-the-year.njk [hunk 1]
```diff
@@ -0,0 +1,57 @@
+{% extends 'layout.njk' %}
+
+{% from 'govuk/components/back-link/macro.njk'     import govukBackLink %}
```
Comment (akaJSaunders) [inline src/views/select-the-year.njk:3]:
```suggestion
{% from 'govuk/components/back-link/macro.njk' import govukBackLink %}
```

File: src/views/select-the-year.njk [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+{% from 'govuk/components/radios/macro.njk' import govukRadios %}
+
+{% block pageTitle %}
+  Select your penalty details
+{% endblock %}
+
+{% block backLink %}
+  {{ 
+    govukBackLink({
+      text: "Back",
```
Comment (akaJSaunders) [inline src/views/select-the-year.njk:14]:
```suggestion
      text: 'Back',
```

File: src/controllers/PenaltyDetailsController.ts [hunk 3]
```diff
@@ -19,7 +19,7 @@ const navigation = {
         return ROOT_URI;
     },
     next(): string {
-        return REVIEW_PENALTY_PAGE_URI;
+        return SELECT_YEAR_PAGE_URI;
```
Comment (akaJSaunders) [inline src/controllers/PenaltyDetailsController.ts:22]: ReviewPenaltyController.ts navigation needs to be updated due to this change

File: src/controllers/SelectYearController.ts [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+                                    ...that.prepareNavigationConfig()
+                                }
+                            );
+                        }
+                    }
+
+                    if (that.formSanitizeFunction != null) {
+                        request.body = that.formSanitizeFunction(request.body);
+                        loggerInstance().debug(`${BaseController.name} - sanitized form body: ${JSON.stringify(request.body)}`);
+                    }
```
Comment (akaJSaunders) [inline src/controllers/SelectYearController.ts:119]: I dont think you need these in your custom action handler

File: src/controllers/SelectYearController.ts [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+        const penaltyList: PenaltyList | undefined = appeal.penaltyIdentifier.penaltyList;
+
+        if (!penaltyList || !penaltyList.items) {
+            throw new Error(SelectYearController.PENALTY_EXPECTED_ERROR);
+        }
+
+        return { penalties: penaltyList.items.map(createPenaltyRadioButton) };
+    }
+
+    protected getExtraActionHandlers(): Record<string, FormActionHandler | FormActionHandlerConstructor> {
```
Comment (akaJSaunders) [inline src/controllers/SelectYearController.ts:96]: Not sure I understand why you needed a custom action handler

File: src/utils/Paths.ts [hunk 6]
```diff
@@ -10,6 +10,7 @@ export const HEALTH_CHECK_URI = `${ROOT_URI}/healthcheck`;
 // Pages
 export const ENTRY_PAGE_URI = `${ROOT_URI}/start`;
 export const PENALTY_DETAILS_PAGE_URI = `${ROOT_URI}/penalty-reference`;
+export const SELECT_YEAR_PAGE_URI = `${ROOT_URI}/select-the-year`;
```
Comment (n-r-b) [inline src/utils/Paths.ts:13]: should the path be `/select-the-penalty`?

File: src/controllers/SelectYearController.ts (hunk 5 above)
Comment (franciscog-kainos) [reply src/controllers/SelectYearController.ts:96]: image

Base Controller:
image

File: src/utils/Paths.ts (hunk 6 above)
Comment (franciscog-kainos) [reply src/utils/Paths.ts:13]: yes oops

File: src/controllers/SelectYearController.ts (hunk 4 above)
Comment (akaJSaunders) [inline src/controllers/SelectYearController.ts:119]: Are you doing any form of santization? I think we cant remove this

File: src/controllers/SelectYearController.ts (hunk 5 above)
Comment (franciscog-kainos) [reply src/controllers/SelectYearController.ts:96]: If I don't include the view model when there is an error, the radio buttons will not populate with data because the assumption was made that the request body will contain the necessary data. Unfortunately when you submit the radio button form you only get the data for the selected radio button

File: src/views/select-the-year.njk [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+      href: navigation.back.href
+    }) 
+  }}
+{% endblock %}
+
+{% block content %}
+
```
Comment (akaJSaunders) [inline src/views/select-the-year.njk:21]:
```suggestion
```

File: src/utils/Paths.ts (hunk 6 above)
Comment (n-r-b) [reply src/utils/Paths.ts:13]: please also update the const name

File: src/controllers/SelectYearController.ts [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+const template = 'select-the-year';
+
+const navigation = {
+    previous(): string {
+        return PENALTY_DETAILS_PAGE_URI;
+    },
+    next(): string {
+        return REVIEW_PENALTY_PAGE_URI;
+    },
+    actions: (changeMode: boolean) => {
```
Comment (n-r-b) [inline src/controllers/SelectYearController.ts:33]: Currently there is no option to change the penalty details from the summary screen, is this needed?

File: src/controllers/SelectYearController.ts [hunk 9]
```diff
@@ -0,0 +1,150 @@
...
+import { PenaltyReferenceRouter } from 'app/middleware/PenaltyReferenceRouter';
+import { Appeal } from 'app/models/Appeal';
+import { ApplicationData, APPLICATION_DATA_KEY } from 'app/models/ApplicationData';
+import { createPenaltyRadioButton } from 'app/models/components/PenaltyRadioButton';
+import { createSchema } from 'app/models/fields/PenaltyChoice.schema';
+import { APPLICATION_DATA_UNDEFINED, SESSION_NOT_FOUND_ERROR } from 'app/utils/CommonErrors';
+import { PENALTY_DETAILS_PAGE_URI, REVIEW_PENALTY_PAGE_URI, SELECT_YEAR_PAGE_URI } from 'app/utils/Paths';
+import { ValidationResult } from 'app/utils/validation/ValidationResult';
+
+const template = 'select-the-year';
```
Comment (n-r-b) [inline src/controllers/SelectYearController.ts:24]: please rename this to `select-the-penalty`

File: src/controllers/SelectYearController.ts (hunk 8 above)
Comment (franciscog-kainos) [reply src/controllers/SelectYearController.ts:33]: image

No you're right however, I used the actions field for consistency as it is used in other action yielding controllers. I can make it indifferent to the change mode

File: src/views/select-the-year.njk [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+  <div class="govuk-width-container">
+    <main role="main" class="govuk-main-wrapper">
+      <div class="govuk-grid-row">
+        <div class="govuk-grid-column-two-thirds">
+          <form action="?{{ navigation.actions.continue }}" method="post" id="penaltyForm">
+
+            {{ 
+              govukRadios({
+                idPrefix: 'select-penalty',
+                name: 'selectPenalty',
```
Comment (n-r-b) [inline src/views/select-the-year.njk:31]: this could also be `'select-penalty'`

File: src/controllers/SelectYearController.ts [hunk 11]
```diff
@@ -0,0 +1,150 @@
...
+            throw SESSION_NOT_FOUND_ERROR;
+        }
+
+        const appData: ApplicationData | undefined = session.getExtraData(APPLICATION_DATA_KEY);
+
+        if(!appData) {
+            throw APPLICATION_DATA_UNDEFINED;
+        }
+
+        const penaltySelected: string = request.body.selectPenalty;
```
Comment (n-r-b) [inline src/controllers/SelectYearController.ts:56]: this variable is redundant, could just do:
`appData.appeal.penaltyIdentifier.penaltyReference = request.body.selectPenalty;`

File: src/controllers/SelectYearController.ts [hunk 12]
```diff
@@ -0,0 +1,150 @@
...
+
+        appData.appeal.penaltyIdentifier.penaltyReference = penaltySelected;
+
+        session.setExtraData(APPLICATION_DATA_KEY, appData);
+    }
+
+}
+
+export const errorMessage: string = 'Select the penalty you want to appeal';
```
Comment (n-r-b) [inline src/controllers/SelectYearController.ts:65]: why not just add this to the `PenaltyChoice.schema.ts` since it is not being re-used elsewhere?

File: src/controllers/SelectYearController.ts [hunk 13]
```diff
@@ -0,0 +1,150 @@
...
+@controller(SELECT_YEAR_PAGE_URI, SessionMiddleware, AuthMiddleware, PenaltyReferenceRouter)
+export class SelectYearController extends SafeNavigationBaseController<any> {
+
+    public static PENALTY_EXPECTED_ERROR: string = 'Penalty object expected but none found';
+
+    constructor() {
+        super(
+            template,
+            navigation,
+            new FormValidator(Joi.object({
```
Comment (n-r-b) [inline src/controllers/SelectYearController.ts:77]: i feel it would be cleaner if we extract the form schema into it's own function

File: src/views/select-the-year.njk (hunk 10 above)
Comment (franciscog-kainos) [reply src/views/select-the-year.njk:31]: The name has to stay as is because that's the key name that gets returned on submit. 'select-penalty' is not a good key name in TS

File: src/models/components/PenaltyRadioButton.ts [hunk 14]
```diff
@@ -0,0 +1,42 @@
...
+    label: {
+        classes: 'govuk-label--s';
+    },
+    hint: {
+        html: string;
+    };
+
+};
+
+export type PenaltyDetailsRadioComponent = {
```
Comment (n-r-b) [inline src/models/components/PenaltyRadioButton.ts:15]: i don't think this is used

File: src/controllers/SelectYearController.ts (hunk 5 above)
Comment (akaJSaunders) [reply src/controllers/SelectYearController.ts:88]: thanks for the explanation

File: test/middleware/PenaltyReferenceRouter.test.ts [hunk 15]
```diff
@@ -0,0 +1,184 @@
...
+import { createSession } from '../utils/session/SessionFactory';
+
+import { PenaltyReferenceRouter } from 'app/middleware/PenaltyReferenceRouter';
+import { ApplicationData, APPLICATION_DATA_KEY } from 'app/models/ApplicationData';
+import { Reasons } from 'app/models/Reasons';
+import { APPLICATION_DATA_UNDEFINED, SESSION_NOT_FOUND_ERROR } from 'app/utils/CommonErrors';
+import { REVIEW_PENALTY_PAGE_URI } from 'app/utils/Paths';
+
+
+
```
Comment (akaJSaunders) [inline test/middleware/PenaltyReferenceRouter.test.ts:14]:
```suggestion
```

Comment (akaJSaunders) [inline test/middleware/PenaltyReferenceRouter.test.ts:14]:
```suggestion

```

File: test/middleware/PenaltyReferenceRouter.test.ts [hunk 16]
```diff
@@ -0,0 +1,184 @@
...
+import { createSubstituteOf } from '../SubstituteFactory';
+import { createSession } from '../utils/session/SessionFactory';
+
+import { PenaltyReferenceRouter } from 'app/middleware/PenaltyReferenceRouter';
+import { ApplicationData, APPLICATION_DATA_KEY } from 'app/models/ApplicationData';
+import { Reasons } from 'app/models/Reasons';
+import { APPLICATION_DATA_UNDEFINED, SESSION_NOT_FOUND_ERROR } from 'app/utils/CommonErrors';
+import { REVIEW_PENALTY_PAGE_URI } from 'app/utils/Paths';
+
+
```
Comment (akaJSaunders) [inline test/middleware/PenaltyReferenceRouter.test.ts:13]: No need for this large gap between tests and imports, also I noticed that you havnt put empty lines between tests, probably worth doing so.

File: src/controllers/SelectYearController.ts [hunk 17]
```diff
@@ -0,0 +1,137 @@
...
+                handle: async (request: Request, response: Response) => {
+                    if (that.validator != null) {
+                        const validationResult: ValidationResult = await that.validator.validate(request);
+                        if (validationResult.errors.length > 0) {
+                            return await that.renderWithStatus(UNPROCESSABLE_ENTITY)(
+                                that.template,
+                                {
+                                    ...request.body,
+                                    validationResult,
+                                    ...that.prepareViewModel(),
```
Comment (n-r-b) [inline src/controllers/SelectYearController.ts:101]: I'm wondering if other controllers would benefit from adding this line to the Base Controller's getDefaultActionHandler().. or can we not just add a custom validator that retains this data in some way?

File: src/controllers/SelectYearController.ts [hunk 18]
```diff
@@ -0,0 +1,137 @@
...
+        session.setExtraData(APPLICATION_DATA_KEY, appData);
+    }
+
+}
+
+// tslint:disable-next-line: max-classes-per-file
+@controller(SELECT_THE_PENALTY_PAGE_URI, SessionMiddleware, AuthMiddleware, PenaltyReferenceRouter)
+export class SelectYearController extends SafeNavigationBaseController<any> {
+
+    public static PENALTY_EXPECTED_ERROR: string = 'Penalty object expected but none found';
```
Comment (n-r-b) [inline src/controllers/SelectYearController.ts:65]: can this be `private`?

File: src/controllers/SelectYearController.ts [hunk 19]
```diff
@@ -0,0 +1,137 @@
...
+        appData.appeal.penaltyIdentifier.penaltyReference = request.body.selectPenalty;
+
+        session.setExtraData(APPLICATION_DATA_KEY, appData);
+    }
+
+}
+
+// tslint:disable-next-line: max-classes-per-file
+@controller(SELECT_THE_PENALTY_PAGE_URI, SessionMiddleware, AuthMiddleware, PenaltyReferenceRouter)
+export class SelectYearController extends SafeNavigationBaseController<any> {
```
Comment (n-r-b) [inline src/controllers/SelectYearController.ts:63]: Can we rename this class to `SelectPenaltyController`?

File: src/controllers/SelectYearController.ts (hunk 17 above)
Comment (franciscog-kainos) [reply src/controllers/SelectYearController.ts:101]: It's probably not a safe bet as some keys in the validation result can be the same as in the view model. Custom validator wouldn't really make it simpler as I would have to reconstruct the radio buttons in it

Comment (franciscog-kainos) [reply src/controllers/SelectYearController.ts:101]: If the default action handler was comprised of made of smaller reusable actions *hint* then this wouldn't be a problem

File: src/middleware/PenaltyReferenceRouter.ts [hunk 20]
```diff
@@ -0,0 +1,50 @@
...
+            .appeal
+            .penaltyIdentifier
+            .penaltyList?.items
+            .length;
+
+        if (!numberOfPenalties) {
+            throw PenaltyReferenceRouter.PENALTY_LIST_UNDEFINED_ERROR;
+        }
+
+        if (numberOfPenalties === 1) {
```
Comment (n-r-b) [inline src/middleware/PenaltyReferenceRouter.ts:38]: can `numberOfPenalties` be 0?

File: src/models/PenaltyIdentifier.ts [hunk 21]
```diff
@@ -2,7 +2,8 @@ import { PenaltyList } from 'ch-sdk-node/dist/services/lfp';
 
 export interface PenaltyIdentifier {
     companyNumber: string;
-    penaltyReference: string;
+    userInputPenaltyReference: string;
+    penaltyReference?: string;
```
Comment (n-r-b) [inline src/models/PenaltyIdentifier.ts:6]: should this be a required field? We could set this to the value of the `userInputPenaltyReference` so that it's never null

Comment (franciscog-kainos) [reply src/models/PenaltyIdentifier.ts:6]: It's not that it isn't required, it's that it won't be defined when the object is instantiated

File: src/middleware/PenaltyReferenceRouter.ts (hunk 20 above)
Comment (franciscog-kainos) [reply src/middleware/PenaltyReferenceRouter.ts:38]: No but I'll change that to `<`

File: src/models/PenaltyIdentifier.ts (hunk 21 above)
Comment (franciscog-kainos) [reply src/models/PenaltyIdentifier.ts:6]: There's some some logic behind this which does that after the user has pressed continue in the PenaltyDetailsPage

File: test/controllers/processors/AppealStorageFormSubmissionProcessor.test.ts [hunk 22]
```diff
@@ -24,7 +24,7 @@ describe('AppealStorageForSubmissionProcessor', () => {
     const appeal: Appeal = {
         penaltyIdentifier: {
             companyNumber: '00345567',
-            penaltyReference: 'A00000001',
+            userInputPenaltyReference: 'A00000001',
```
Comment (n-r-b) [inline test/controllers/processors/AppealStorageFormSubmissionProcessor.test.ts:27]: what's the reason for changing this? only the `penaltyReference` is saved in mongo?

File: src/controllers/ReviewPenaltyController.ts [hunk 23]
```diff
@@ -23,22 +27,44 @@ const navigation = {
...
+        const penaltyList = appeal.penaltyIdentifier.penaltyList;
+
+        if (!penaltyList) {
             throw new Error(ReviewPenaltyController.PENALTY_EXPECTED_ERROR);
+        }
+
+        let penaltyReference: string | undefined;
+
+        if (penaltyList.items.length === 1) {
+            penaltyReference = appeal.penaltyIdentifier.userInputPenaltyReference;
```
Comment (n-r-b) [inline src/controllers/ReviewPenaltyController.ts:48]: if there is only one item in the penaltyList, at which point does the `appeal.penaltyIdentifier.penaltyReference` get set?

Comment (franciscog-kainos) [reply src/controllers/ReviewPenaltyController.ts:48]: In the processor of the SelectPenaltyController

File: test/controllers/processors/AppealStorageFormSubmissionProcessor.test.ts (hunk 22 above)
Comment (franciscog-kainos) [reply test/controllers/processors/AppealStorageFormSubmissionProcessor.test.ts:27]: My bad, didn't pay attention

File: src/controllers/SelectYearController.ts (hunk 17 above)
Comment (n-r-b) [reply src/controllers/SelectYearController.ts]: I feel it would be cleaner if we created a custom validator over duplicating the default action handler, if it's not too much effort? We can reuse the logic we have in place to construct the radio buttons

File: src/controllers/ReviewPenaltyController.ts (hunk 23 above)
Comment (n-r-b) [reply src/controllers/ReviewPenaltyController.ts:48]: what happens if the `PenaltyReferenceRouter` has redirected us to this page from the penalty details page?

Comment (franciscog-kainos) [reply src/controllers/ReviewPenaltyController.ts:48]: then it must mean that there is only one penalty reference and it will show you the table as it should? What are you asking?

Comment (n-r-b) [reply src/controllers/ReviewPenaltyController.ts:48]: i was just double checking that the `penaltyReference` is still stored in the session, as before
